bridge/src/message_to_mainnet.rs

```diff
@@ -0,0 +1,109 @@
+use ethereum_types::{Address, U256, H256};
+use contracts::foreign::events::Withdraw;
+use web3::types::Log;
+use ethabi;
+use error::{Error, ErrorKind};
+
+/// the message that is relayed from side to main.
+/// contains all the information required for the relay.
+/// validators sign off on this message.
+#[derive(PartialEq, Debug)]
+pub struct MessageToMainnet {
+	pub recipient: Address,
+	pub value: U256,
+	pub sidenet_transaction_hash: H256,
+	pub mainnet_gas_price: U256,
+}
+
+/// length of a `MessageToMainnet.to_bytes()` in bytes
+pub const MESSAGE_LENGTH: usize = 116;
+
+impl MessageToMainnet {
+	/// parses message from a byte slice
+	pub fn from_bytes(bytes: &[u8]) -> Self {
+		assert_eq!(bytes.len(), MESSAGE_LENGTH);
+
+		Self {
+			recipient: bytes[0..20].into(),
+			value: U256::from_little_endian(&bytes[20..52]),
+			sidenet_transaction_hash: bytes[52..84].into(),
+			mainnet_gas_price: U256::from_little_endian(&bytes[84..MESSAGE_LENGTH]),
+		}
+	}
+
+	/// construct a message from a `Withdraw` event that was logged on `foreign`
+	pub fn from_log(web3_log: Log) -> Result<Self, Error> {
+		let ethabi_raw_log = ethabi::RawLog {
+			topics: web3_log.topics,
+			data: web3_log.data.0,
+		};
+		let withdraw_log = Withdraw::default().parse_log(ethabi_raw_log)?;
+		let hash = web3_log.transaction_hash
+			.ok_or(Error::from_kind(ErrorKind::Msg("`log` must be mined and contain `transaction_hash`".into())))?;
+		Ok(Self {
+			recipient: withdraw_log.recipient.0.into(),
+			value: U256(withdraw_log.value.0),
+			sidenet_transaction_hash: hash.to_vec().as_slice().into(),
+			mainnet_gas_price: U256(withdraw_log.home_gas_price.0),
+		})
+	}
+
+	/// serializes message to a byte vector.
+	/// mainly used to construct the message byte vector that is then signed
+	/// and passed to `ForeignBridge.submitSignature`
+	pub fn to_bytes(&self) -> Vec<u8> {
+		let mut result = vec![0u8; MESSAGE_LENGTH];
+		result[0..20].copy_from_slice(&self.recipient.0[..]);
+		self.value.to_little_endian(&mut result[20..52]);
+		result[52..84].copy_from_slice(&self.sidenet_transaction_hash.0[..]);
+		self.mainnet_gas_price.to_little_endian(&mut result[84..MESSAGE_LENGTH]);
+		return result;
+	}
+
+	/// serializes message to an ethabi payload
+	pub fn to_payload(&self) -> Vec<u8> {
+		ethabi::encode(&[ethabi::Token::Bytes(self.to_bytes())])
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use quickcheck::TestResult;
+	use super::*;
+
+	quickcheck! {
+		fn quickcheck_message_to_mainnet_roundtrips_to_bytes(
+			recipient_raw: Vec<u8>,
+			value_raw: u64,
+			sidenet_transaction_hash_raw: Vec<u8>,
+			mainnet_gas_price_raw: u64
+		) -> TestResult {
+			if recipient_raw.len() != 20 || sidenet_transaction_hash_raw.len() != 32 {
+				return TestResult::discard();
```

Comment:
is it possible to specify the length of these vectors? If not, the chance that both of them will have desired len is pretty low...

Reply:
it's possible but more work. one would have to create a new type that implements quickchecks `Arbitrary` and only generates vectors of a specific size.

quickcheck by default will fail if it doesn't generate at least 100 tests.
so we're guaranteed that it generates at least 100 vectors of the desired lengths.

this is not perfect but good enough in my opinion.

> N.B. Since discarding a test means it neither passes nor fails, quickcheck will try to replace the discarded test with a fresh one. However, if your condition is seldom met, it's possible that quickcheck will have to settle for running fewer tests than usual. By default, if quickcheck can't find 100 valid tests after trying 10,000 times, then it will give up. 

https://github.com/BurntSushi/quickcheck#discarding-test-results-or-properties-are-polymorphic

Reply:
interesting. let's leave it as it is for now :)

Reply:
now i remember the full context. this is just a workaround. the best solution would be to implement `quickcheck::Arbitrary` for the types in ethereum-types! ethereum-types could benefit from some quickchecking itself as well. i'll give it a shot next week.

Reply:
yes. and let's hide it behind feature gate :)

Reply:
added https://github.com/paritytech/primitives/issues/21